quill-core/src/main/scala/io/getquill/norm/BetaReduction.scala

```diff
@@ -17,11 +17,14 @@ case class BetaReduction(map: collection.Map[Ident, Ast])
       case FunctionApply(Function(params, body), values) =>
         apply(BetaReduction(map ++ params.zip(values)).apply(body))
       case ident: Ident =>
-        map.getOrElse(ident, ident)
+        map.get(ident).map(BetaReduction(map - ident)(_)).getOrElse(ident)
       case Function(params, body) =>
         Function(params, BetaReduction(map -- params)(body))
       case OptionOperation(t, a, b, c) =>
         OptionOperation(t, apply(a), b, BetaReduction(map - b)(c))
+      case Block(statements) =>
+        val vals = statements.collect { case x: Val => x.name -> x.body }
```

Comment:
@fwbrasil I added a test case to cover nested block usage. 

Scoping should only be an issue when the user shadows previously defined val(s) from an outer scope. Of course this is possible but I don't see a general solution to the problem given that the traversed stack (i.e. `Map[Ident, Ast]`) can only store unique `Ident`s, and when we pop an ident off the stack there's then no longer a way to reduce the shadowed val.